Readme.md
# koba

A service is a set of functions. This module serves them over HTTP without
asking the functions to know it: an API is an ordinary Go value, and every
exported method of that value is an endpoint.

```go
type Polls struct{}

func (*Polls) ListPolls() []Poll { ... }

func main() {
	app := koba.MakeRawApp()
	app.AddApi(&Polls{})
	log.Fatal(app.Run(":8080"))
}
```

`ListPolls` is served at `GET /list-polls` and answers with the JSON
encoding of what it returns. A method that takes a value is served over
`POST`, and the value is decoded from the JSON body:

```go
func (*Polls) Cast(request CastRequest) Receipt { ... }
```

## Conventions

* Every exported method of a registered value is an endpoint. A method takes
  at most one parameter, the request, and returns at most one value, the
  response; any other shape cannot be served, and registering it is an error.
* The HTTP method follows the signature: a method that takes a request is
  served over `POST`, and the request is decoded from the JSON body; a method
  that takes nothing is served over `GET`.
* The route is the method's name in lower case, with a dash between words:
  `DeclarePoll` is served at `/declare-poll`. An API that implements
  `WithRoutesAPI` chooses the routes of the methods it names; `Routes` itself
  is not an endpoint.
* A method that returns a value answers `200 OK` with the value encoded as
  JSON; a method that returns nothing answers `204 No Content`.
* A body that does not decode into the request answers `400 Bad Request`
  with the same kind of JSON body as a failure.
* A method that panics answers `500 Internal Server Error` with a JSON body
  that carries the message: `{"status": 500, "error":
  "internal_server_error", "message": "..."}`.

## The router

The endpoints are served by `internal/httprouter`, a small router over
`net/http`'s `ServeMux`: method-aware patterns, path values, a middleware
chain and a request context. An app exposes it through `WithRouter` for the
rare need this module does not cover.


CHANGELOG.md
# Changelog

## 0.1.0 (2022-01-12)

### Features

* a function is an endpoint: an API is a value whose exported methods are its operations, and the signature decides how each is called


app_test.go
package koba

import (
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

type greetings struct {
	calls int
}

func (api *greetings) Hello() map[string]string {
	api.calls++
	return map[string]string{"hello": "world"}
}

func (*greetings) Nothing() {}

func (*greetings) Fail() {
	panic(errors.New("the greeting failed"))
}

func (*greetings) FailWithValue() {
	panic(42)
}

func (*greetings) ListHTTPStatuses() []int {
	return []int{http.StatusOK}
}

type renamed struct{}

func (*renamed) Hello() string { return "hi" }

func (*renamed) Other() string { return "other" }

func (*renamed) Routes() Routes {
	return Routes{"Hello": "/greeting"}
}

type polls struct {
	accepted []ballot
}

type ballot struct {
	Nullifier string `json:"nullifier"`
	Sealed    string `json:"sealed"`
}

func (api *polls) Cast(request ballot) ballot {
	api.accepted = append(api.accepted, request)
	return request
}

func (api *polls) Count() int {
	return len(api.accepted)
}

func (*polls) Forget(ballot) {}

type takesTwo struct{}

func (*takesTwo) Do(int, int) {}

type takesPointer struct{}

func (*takesPointer) Do(*ballot) {}

type returnsTwo struct{}

func (*returnsTwo) Do() (int, int) { return 1, 2 }

func call(t *testing.T, served App, method, target string) *httptest.ResponseRecorder {
	t.Helper()
	return send(t, served, method, target, "")
}

func send(t *testing.T, served App, method, target, body string) *httptest.ResponseRecorder {
	t.Helper()
	response := httptest.NewRecorder()
	served.(WithRouter).Router().ServeHTTP(response, httptest.NewRequest(method, target, strings.NewReader(body)))
	return response
}

func TestTheHTTPMethodFollowsTheSignature(t *testing.T) {
	api := &polls{}
	served := MakeRawApp()
	served.AddApi(api)

	response := send(t, served, http.MethodPost, "/cast", `{"nullifier":"n1","sealed":"s1"}`)
	if response.Code != http.StatusOK || response.Body.String() != `{"nullifier":"n1","sealed":"s1"}` {
		t.Fatalf("POST /cast = %d %q", response.Code, response.Body.String())
	}
	if len(api.accepted) != 1 || api.accepted[0].Nullifier != "n1" {
		t.Fatalf("accepted = %+v", api.accepted)
	}
	if response := call(t, served, http.MethodGet, "/count"); response.Code != http.StatusOK || response.Body.String() != "1" {
		t.Fatalf("GET /count = %d %q", response.Code, response.Body.String())
	}
	if response := send(t, served, http.MethodPost, "/forget", `{}`); response.Code != http.StatusNoContent {
		t.Fatalf("POST /forget = %d", response.Code)
	}
	if response := call(t, served, http.MethodGet, "/cast"); response.Code != http.StatusNotFound {
		t.Fatalf("GET /cast = %d, want 404: a method with a request is not served over GET", response.Code)
	}
	if response := send(t, served, http.MethodPost, "/count", "{}"); response.Code != http.StatusNotFound {
		t.Fatalf("POST /count = %d, want 404: a method without a request is not served over POST", response.Code)
	}
}

func TestABodyThatDoesNotDecodeIsRejected(t *testing.T) {
	served := MakeRawApp()
	served.AddApi(&polls{})
	for name, body := range map[string]string{"malformed": `{"nullifier":`, "empty": "", "wrong type": `{"nullifier":2}`} {
		t.Run(name, func(t *testing.T) {
			response := send(t, served, http.MethodPost, "/cast", body)
			if response.Code != http.StatusBadRequest {
				t.Fatalf("status = %d, want 400: %s", response.Code, response.Body.String())
			}
			var decoded map[string]any
			if err := json.Unmarshal(response.Body.Bytes(), &decoded); err != nil {
				t.Fatal(err)
			}
			if decoded["error"] != "bad_request" || !strings.HasPrefix(decoded["message"].(string), "request body: ") {
				t.Fatalf("body = %v", decoded)
			}
		})
	}
}

func TestExportedMethodsAreEndpoints(t *testing.T) {
	api := &greetings{}
	served := MakeRawApp()
	served.AddApi(api)

	response := call(t, served, http.MethodGet, "/hello")
	if response.Code != http.StatusOK || response.Body.String() != `{"hello":"world"}` {
		t.Fatalf("GET /hello = %d %q", response.Code, response.Body.String())
	}
	if got := response.Header().Get("Content-Type"); !strings.HasPrefix(got, "application/json") {
		t.Fatalf("Content-Type = %q", got)
	}
	if api.calls != 1 {
		t.Fatalf("calls = %d, want 1", api.calls)
	}
	if response := call(t, served, http.MethodGet, "/nothing"); response.Code != http.StatusNoContent || response.Body.Len() != 0 {
		t.Fatalf("GET /nothing = %d %q", response.Code, response.Body.String())
	}
	if response := call(t, served, http.MethodGet, "/list-http-statuses"); response.Code != http.StatusOK || response.Body.String() != "[200]" {
		t.Fatalf("GET /list-http-statuses = %d %q", response.Code, response.Body.String())
	}
	if response := call(t, served, http.MethodGet, "/missing"); response.Code != http.StatusNotFound {
		t.Fatalf("GET /missing = %d", response.Code)
	}
}

func TestAFailingMethodIsAnsweredWithTheFailure(t *testing.T) {
	served := MakeRawApp()
	served.AddApi(&greetings{})
	for target, message := range map[string]string{"/fail": "the greeting failed", "/fail-with-value": "42"} {
		response := call(t, served, http.MethodGet, target)
		if response.Code != http.StatusInternalServerError {
			t.Fatalf("GET %s = %d", target, response.Code)
		}
		var body map[string]any
		if err := json.Unmarshal(response.Body.Bytes(), &body); err != nil {
			t.Fatal(err)
		}
		if body["status"] != float64(500) || body["error"] != "internal_server_error" || body["message"] != message {
			t.Fatalf("GET %s body = %v", target, body)
		}
	}
}

func TestAnAPIChoosesItsRoutes(t *testing.T) {
	served := MakeRawApp()
	served.AddApi(&renamed{})
	if response := call(t, served, http.MethodGet, "/greeting"); response.Code != http.StatusOK || response.Body.String() != `"hi"` {
		t.Fatalf("GET /greeting = %d %q", response.Code, response.Body.String())
	}
	if response := call(t, served, http.MethodGet, "/other"); response.Code != http.StatusOK {
		t.Fatalf("GET /other = %d", response.Code)
	}
	for _, target := range []string{"/hello", "/routes"} {
		if response := call(t, served, http.MethodGet, target); response.Code != http.StatusNotFound {
			t.Fatalf("GET %s = %d, want 404", target, response.Code)
		}
	}
}

func TestRegistrationRefusesShapesNoEndpointCanServe(t *testing.T) {
	for name, api := range map[string]any{"two parameters": &takesTwo{}, "pointer parameter": &takesPointer{}, "two results": &returnsTwo{}, "nil": nil} {
		t.Run(name, func(t *testing.T) {
			defer func() {
				if recover() == nil {
					t.Fatal("AddApi did not panic")
				}
			}()
			MakeRawApp().AddApi(api)
		})
	}
}

func TestRouteOf(t *testing.T) {
	for name, want := range map[string]string{
		"Hello":            "/hello",
		"DeclarePoll":      "/declare-poll",
		"ListHTTPStatuses": "/list-http-statuses",
		"HTTP":             "/http",
		"Item2Get":         "/item2-get",
		"A":                "/a",
	} {
		if got := RouteOf(name); got != want {
			t.Errorf("RouteOf(%q) = %q, want %q", name, got, want)
		}
	}
}


internal/httprouter/router_test.go
package httprouter

import (
	"bytes"
	"context"
	"errors"
	"io"
	"net"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func serve(t *testing.T, handler http.Handler, method, target string, body io.Reader) *httptest.ResponseRecorder {
	t.Helper()
	response := httptest.NewRecorder()
	handler.ServeHTTP(response, httptest.NewRequest(method, target, body))
	return response
}

func TestRoutesMatchByMethodAndPattern(t *testing.T) {
	engine := New()
	engine.GET("/items/{id}", func(c *Context) { c.String(http.StatusOK, "get %s", c.Param("id")) })
	engine.POST("/items", func(c *Context) { c.Status(http.StatusCreated) })
	engine.PUT("/items/:id", func(c *Context) { c.String(http.StatusOK, "put %s", c.Params.ByName("id")) })
	engine.PATCH("/items/{id}", func(c *Context) { c.String(http.StatusOK, "patch") })
	engine.DELETE("/items/{id}", func(c *Context) { c.Status(http.StatusNoContent) })
	engine.GET("/items/special", func(c *Context) { c.String(http.StatusOK, "special") })
	engine.GET("/files/*rest", func(c *Context) { c.String(http.StatusOK, "file %s", c.Param("rest")) })
	engine.GET("/", func(c *Context) { c.String(http.StatusOK, "root") })

	for _, test := range []struct {
		method, target string
		wantStatus     int
		wantBody       string
	}{
		{http.MethodGet, "/items/item-1", http.StatusOK, "get item-1"},
		{http.MethodPost, "/items", http.StatusCreated, ""},
		{http.MethodPut, "/items/item-2", http.StatusOK, "put item-2"},
		{http.MethodPatch, "/items/item-3", http.StatusOK, "patch"},
		{http.MethodDelete, "/items/item-4", http.StatusNoContent, ""},
		{http.MethodGet, "/items/special", http.StatusOK, "special"},
		{http.MethodGet, "/files/a/b/c.txt", http.StatusOK, "file a/b/c.txt"},
		{http.MethodGet, "/files/", http.StatusOK, "file "},
		{http.MethodGet, "/", http.StatusOK, "root"},
		{http.MethodGet, "/missing", http.StatusNotFound, "404 page not found"},
		{http.MethodDelete, "/items", http.StatusNotFound, "404 page not found"},
	} {
		response := serve(t, engine, test.method, test.target, nil)
		if response.Code != test.wantStatus || response.Body.String() != test.wantBody {
			t.Errorf("%s %s = %d %q, want %d %q", test.method, test.target, response.Code, response.Body.String(), test.wantStatus, test.wantBody)
		}
	}
}

func TestPatternsAreNormalized(t *testing.T) {
	for input, want := range map[string]string{
		"/items/:id":       "/items/{id}",
		"/items/{id}/sub":  "/items/{id}/sub",
		"/docs/*asset":     "/docs/{asset...}",
		"/docs/{asset...}": "/docs/{asset...}",
		"/":                "/{$}",
		"/items/":          "/items/{$}",
		"/a/:x/b/:y":       "/a/{x}/b/{y}",
		"/exact/{$}":       "/exact/{$}",
	} {
		if got := toPattern(input).text; got != want {
			t.Errorf("toPattern(%q) = %q, want %q", input, got, want)
		}
	}
	if got := toPattern("/a/:x/b/*y").names; strings.Join(got, ",") != "x,y" {
		t.Errorf("names = %v, want [x y]", got)
	}
}

func TestGroupsJoinPrefixesAndPrependMiddleware(t *testing.T) {
	engine := New()
	var order []string
	engine.Use(func(c *Context) { order = append(order, "engine") })
	service := engine.Group("/service", func(c *Context) { order = append(order, "service") })
	if got := service.BasePath(); got != "/service" {
		t.Fatalf("BasePath() = %q, want /service", got)
	}
	items := service.Group("/items/:id")
	items.Use(func(c *Context) { order = append(order, "items") })
	items.GET("", func(c *Context) {
		order = append(order, "handler")
		c.String(http.StatusOK, c.FullPath()+" "+c.Param("id"))
	})
	engine.Use(func(c *Context) { order = append(order, "late") })

	response := serve(t, engine, http.MethodGet, "/service/items/item-1", nil)
	if got, want := response.Body.String(), "/service/items/{id} item-1"; got != want {
		t.Fatalf("body = %q, want %q", got, want)
	}
	if got, want := strings.Join(order, ">"), "engine>service>items>handler"; got != want {
		t.Fatalf("order = %q, want %q", got, want)
	}
}

func TestNextAndAbortControlTheChain(t *testing.T) {
	engine := New()
	var order []string
	engine.Use(func(c *Context) {
		order = append(order, "before")
		c.Next()
		order = append(order, "after:"+http.StatusText(c.Writer.Status()))
	})
	engine.Use(func(c *Context) { order = append(order, "returns without Next") })
	engine.GET("/abort", func(c *Context) {
		order = append(order, "abort")
		c.AbortWithStatusJSON(http.StatusForbidden, H{"error": "forbidden"})
	}, func(c *Context) { order = append(order, "never") })
	engine.GET("/ok", func(c *Context) { order = append(order, "ok"); c.Status(http.StatusAccepted) })

	response := serve(t, engine, http.MethodGet, "/abort", nil)
	if response.Code != http.StatusForbidden || response.Body.String() != `{"error":"forbidden"}` {
		t.Fatalf("abort response = %d %q", response.Code, response.Body.String())
	}
	if got, want := strings.Join(order, ">"), "before>returns without Next>abort>after:Forbidden"; got != want {
		t.Fatalf("order = %q, want %q", got, want)
	}
	order = nil
	response = serve(t, engine, http.MethodGet, "/ok", nil)
	if response.Code != http.StatusAccepted {
		t.Fatalf("status = %d, want 202", response.Code)
	}
	if got, want := strings.Join(order, ">"), "before>returns without Next>ok>after:Accepted"; got != want {
		t.Fatalf("order = %q, want %q", got, want)
	}
}

func TestUnmatchedRequestsRunTheEngineMiddlewareWithoutAFullPath(t *testing.T) {
	engine := New()
	var seen []string
	engine.Use(func(c *Context) {
		seen = append(seen, "path="+c.FullPath())
		if c.Request.URL.Path == "/probe" {
			c.String(http.StatusOK, "probe")
			c.Abort()
		}
	})
	engine.NoRoute(func(c *Context) { c.String(http.StatusGone, "gone") })
	engine.GET("/items", func(c *Context) { c.Status(http.StatusNoContent) })

	if response := serve(t, engine, http.MethodGet, "/probe", nil); response.Code != http.StatusOK || response.Body.String() != "probe" {
		t.Fatalf("probe = %d %q", response.Code, response.Body.String())
	}
	if response := serve(t, engine, http.MethodGet, "/other", nil); response.Code != http.StatusGone || response.Body.String() != "gone" {
		t.Fatalf("no-route handler = %d %q", response.Code, response.Body.String())
	}
	if response := serve(t, engine, http.MethodPost, "/items", nil); response.Code != http.StatusGone {
		t.Fatalf("known path under another method = %d, want the no-route chain", response.Code)
	}
	if got, want := strings.Join(seen, ","), "path=,path=,path="; got != want {
		t.Fatalf("full paths = %q, want %q", got, want)
	}
	plain := New()
	plain.Use(func(c *Context) { c.Status(http.StatusTeapot) })
	if response := serve(t, plain, http.MethodGet, "/nowhere", nil); response.Code != http.StatusTeapot || response.Body.Len() != 0 {
		t.Fatalf("status set by middleware = %d %q, want 418 and no body", response.Code, response.Body.String())
	}
}

func TestCustomMethodsAndHeadersAndQuery(t *testing.T) {
	engine := New()
	engine.Handle("BREW", "/coffee", func(c *Context) {
		c.Header("X-Kind", c.Query("kind"))
		c.Header("X-Removed", "value")
		c.Header("X-Removed", "")
		c.String(http.StatusOK, c.GetHeader("X-Cup"))
	})
	response := serve(t, engine, "BREW", "/coffee?kind=espresso", nil)
	if response.Code != http.StatusOK {
		t.Fatalf("status = %d", response.Code)
	}
	if got := response.Header().Get("X-Kind"); got != "espresso" {
		t.Fatalf("X-Kind = %q", got)
	}
	if _, present := response.Header()["X-Removed"]; present {
		t.Fatal("empty Header value did not remove the header")
	}
	request := httptest.NewRequest("BREW", "/coffee", nil)
	request.Header.Set("X-Cup", "large")
	response = httptest.NewRecorder()
	engine.ServeHTTP(response, request)
	if response.Body.String() != "large" {
		t.Fatalf("body = %q, want large", response.Body.String())
	}
}

func TestResponseWriterRecordsTheStatusUntilTheBodyStarts(t *testing.T) {
	recorder := httptest.NewRecorder()
	c, _ := CreateTestContext(recorder)
	if c.Writer.Status() != http.StatusOK || c.Writer.Size() != -1 || c.Writer.Written() {
		t.Fatalf("fresh writer = %d/%d/%v", c.Writer.Status(), c.Writer.Size(), c.Writer.Written())
	}
	c.Status(http.StatusAccepted)
	if recorder.Code != http.StatusOK || c.Writer.Written() {
		t.Fatal("Status sent the headers early")
	}
	if _, err := c.Writer.WriteString("ab"); err != nil {
		t.Fatal(err)
	}
	c.Writer.WriteHeader(http.StatusTeapot)
	if recorder.Code != http.StatusAccepted || c.Writer.Status() != http.StatusAccepted || c.Writer.Size() != 2 {
		t.Fatalf("after write = %d/%d/%d", recorder.Code, c.Writer.Status(), c.Writer.Size())
	}
	c.Writer.Flush()
	if _, _, err := c.Writer.Hijack(); err == nil {
		t.Fatal("Hijack succeeded on a recorder")
	}
	if http.NewResponseController(c.Writer).Flush() != nil {
		t.Fatal("ResponseController cannot reach the recorder")
	}
}

func TestDataJSONStringAndRedirect(t *testing.T) {
	engine := New()
	engine.GET("/json", func(c *Context) { c.JSON(http.StatusCreated, H{"b": 1, "a": "x"}) })
	engine.GET("/typed", func(c *Context) {
		c.Header("Content-Type", "application/problem+json")
		c.JSON(http.StatusBadRequest, H{"error": true})
	})
	engine.GET("/empty", func(c *Context) { c.JSON(http.StatusNoContent, H{"ignored": true}) })
	engine.GET("/data", func(c *Context) { c.Data(http.StatusOK, "text/html", []byte("<b>hi</b>")) })
	engine.GET("/text", func(c *Context) { c.String(http.StatusOK, "100%% %d", 5) })
	engine.GET("/raw", func(c *Context) { c.String(http.StatusOK, "100%") })
	engine.GET("/docs", func(c *Context) { c.Redirect(http.StatusPermanentRedirect, "/docs/") })

	for _, test := range []struct {
		target, wantType, wantBody string
		wantStatus                 int
	}{
		{"/json", "application/json; charset=utf-8", `{"a":"x","b":1}`, http.StatusCreated},
		{"/typed", "application/problem+json", `{"error":true}`, http.StatusBadRequest},
		{"/empty", "application/json; charset=utf-8", "", http.StatusNoContent},
		{"/data", "text/html", "<b>hi</b>", http.StatusOK},
		{"/text", "text/plain; charset=utf-8", "100% 5", http.StatusOK},
		{"/raw", "text/plain; charset=utf-8", "100%", http.StatusOK},
	} {
		response := serve(t, engine, http.MethodGet, test.target, nil)
		if response.Code != test.wantStatus || response.Body.String() != test.wantBody || response.Header().Get("Content-Type") != test.wantType {
			t.Errorf("%s = %d %q %q, want %d %q %q", test.target, response.Code, response.Body.String(), response.Header().Get("Content-Type"), test.wantStatus, test.wantBody, test.wantType)
		}
	}
	response := serve(t, engine, http.MethodGet, "/docs", nil)
	if response.Code != http.StatusPermanentRedirect || response.Header().Get("Location") != "/docs/" {
		t.Fatalf("redirect = %d %q", response.Code, response.Header().Get("Location"))
	}
	defer func() {
		if recover() == nil {
			t.Fatal("Redirect accepted a non-redirect status")
		}
	}()
	c, _ := CreateTestContext(httptest.NewRecorder())
	c.Request = httptest.NewRequest(http.MethodGet, "/", nil)
	c.Redirect(http.StatusOK, "/")
}

type payload struct {
	Name string `json:"name"`
}

func TestBindingDecodesJSONAndCachesTheBody(t *testing.T) {
	c, _ := CreateTestContext(httptest.NewRecorder())
	c.Request = httptest.NewRequest(http.MethodPost, "/", strings.NewReader(`{"name":"one"}`))
	var first, second payload
	if err := c.ShouldBindBodyWith(&first, BindingJSON); err != nil || first.Name != "one" {
		t.Fatalf("first decode = %v %+v", err, first)
	}
	cached, _ := c.Get(BodyBytesKey)
	if !bytes.Equal(cached.([]byte), []byte(`{"name":"one"}`)) {
		t.Fatalf("cached body = %q", cached)
	}
	if err := c.ShouldBindBodyWith(&second, BindingJSON); err != nil || second.Name != "one" {
		t.Fatalf("second decode = %v %+v", err, second)
	}
	if err := c.ShouldBindBodyWith(&second, BindingKind(99)); err == nil {
		t.Fatal("unknown binding kind accepted")
	}

	c, _ = CreateTestContext(httptest.NewRecorder())
	c.Request = httptest.NewRequest(http.MethodPost, "/", strings.NewReader(""))
	if err := c.ShouldBindBodyWith(&first, BindingJSON); !errors.Is(err, io.EOF) {
		t.Fatalf("empty body error = %v, want io.EOF", err)
	}

	c, _ = CreateTestContext(httptest.NewRecorder())
	c.Request = httptest.NewRequest(http.MethodPost, "/", strings.NewReader(`{"name":"two"}`))
	if err := c.ShouldBindJSON(&first); err != nil || first.Name != "two" {
		t.Fatalf("ShouldBindJSON = %v %+v", err, first)
	}
	c, _ = CreateTestContext(httptest.NewRecorder())
	c.Request = httptest.NewRequest(http.MethodPost, "/", strings.NewReader(`{"name":`))
	if err := c.BindJSON(&first); err == nil || c.Writer.Status() != http.StatusBadRequest || !c.IsAborted() {
		t.Fatalf("BindJSON on malformed body = %v, status %d, aborted %v", err, c.Writer.Status(), c.IsAborted())
	}
	if c.Writer.Written() {
		t.Fatal("BindJSON sent the headers")
	}
	c, _ = CreateTestContext(httptest.NewRecorder())
	if err := c.ShouldBindJSON(&first); err == nil {
		t.Fatal("a context without a request decoded a body")
	}
}

func TestKeysAndContextDelegation(t *testing.T) {
	c, _ := CreateTestContext(httptest.NewRecorder())
	if c.Err() != nil || c.Done() != nil || c.Value("missing") != nil {
		t.Fatal("a context without a request is not a background context")
	}
	if _, ok := c.Deadline(); ok {
		t.Fatal("a context without a request has a deadline")
	}
	c.Set("user", "ana")
	if value, exists := c.Get("user"); !exists || value != "ana" || c.MustGet("user") != "ana" || c.Value("user") != "ana" {
		t.Fatal("stored value not returned")
	}
	defer func() {
		if recover() == nil {
			t.Fatal("MustGet did not panic")
		}
	}()
	type key struct{}
	ctx, cancel := context.WithCancel(context.WithValue(context.Background(), key{}, "request"))
	c.Request = httptest.NewRequest(http.MethodGet, "/", nil).WithContext(ctx)
	if c.Value(key{}) != "request" {
		t.Fatal("request context value not reached")
	}
	cancel()
	if !errors.Is(c.Err(), context.Canceled) {
		t.Fatal("cancellation not reached")
	}
	<-c.Done()
	c.MustGet("absent")
}

func TestRecoveryAnswersPanicsAndLoggerWritesALine(t *testing.T) {
	var logs, errs bytes.Buffer
	DefaultWriter, DefaultErrorWriter = &logs, &errs
	defer func() { DefaultWriter, DefaultErrorWriter = io.Discard, io.Discard }()
	engine := Default()
	engine.GET("/panic", func(c *Context) { panic("boom") })
	engine.GET("/fine", func(c *Context) { c.Status(http.StatusNoContent) })
	engine.GET("/pipe", func(c *Context) { panic(errors.New("write: broken pipe")) })

	if response := serve(t, engine, http.MethodGet, "/panic", nil); response.Code != http.StatusInternalServerError {
		t.Fatalf("panic status = %d", response.Code)
	}
	if !strings.Contains(errs.String(), "boom") || !strings.Contains(errs.String(), "router_test.go") {
		t.Fatalf("panic log = %q", errs.String())
	}
	if response := serve(t, engine, http.MethodGet, "/fine?x=1", nil); response.Code != http.StatusNoContent {
		t.Fatalf("status = %d", response.Code)
	}
	if !strings.Contains(logs.String(), "| 500 |") || !strings.Contains(logs.String(), "GET     /fine?x=1") {
		t.Fatalf("log lines = %q", logs.String())
	}
}

func TestWrapHAndStandardHandlers(t *testing.T) {
	engine := New()
	mux := http.NewServeMux()
	mux.HandleFunc("/wrapped/", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusAccepted)
		_, _ = io.WriteString(w, r.URL.Path)
	})
	after := false
	engine.GET("/wrapped/{rest...}", WrapH(mux), func(c *Context) {
		after = c.Writer.Status() == http.StatusAccepted && c.Writer.Size() == len("/wrapped/x")
	})
	response := serve(t, engine, http.MethodGet, "/wrapped/x", nil)
	if response.Code != http.StatusAccepted || response.Body.String() != "/wrapped/x" || !after {
		t.Fatalf("wrapped = %d %q, after=%v", response.Code, response.Body.String(), after)
	}
}

func TestRegistrationRefusesWhatItCannotRoute(t *testing.T) {
	for name, register := range map[string]func(*Engine){
		"no method":  func(e *Engine) { e.Handle("", "/x", func(*Context) {}) },
		"no handler": func(e *Engine) { e.GET("/x") },
		"conflict": func(e *Engine) {
			e.GET("/a/{x}/c", func(*Context) {})
			e.GET("/a/b/{y}", func(*Context) {})
		},
	} {
		t.Run(name, func(t *testing.T) {
			defer func() {
				if recover() == nil {
					t.Fatal("registration did not panic")
				}
			}()
			register(New())
		})
	}
}

func TestRunRefusesABusyAddress(t *testing.T) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Skip(err)
	}
	defer listener.Close()
	if err := New().Run(listener.Addr().String()); err == nil {
		t.Fatal("Run bound an address in use")
	}
}


endpoints.go
package koba

import (
	"errors"
	"fmt"
	"net/http"
	"reflect"
	"strings"
	"unicode"

	"github.com/roxiegrillo/koba/v2/internal/httprouter"
)

// endpoint is one exported method of an API, served at a route.
type endpoint struct {
	name       string
	route      string
	httpMethod string
	method     reflect.Method
	call       reflect.Value
}

// endpointsOf lists the endpoints api's exported methods make, in the order
// reflect reports them. A method is served when it takes at most one
// parameter, the request, and returns at most one value, the response; any
// other shape is an error, so that a method nobody could call is noticed when
// the API is registered.
func endpointsOf(api any) ([]endpoint, error) {
	apiType := reflect.TypeOf(api)
	if apiType == nil {
		return nil, errors.New("koba: an API is a value, not nil")
	}
	var routes Routes
	if chooser, ok := api.(WithRoutesAPI); ok {
		routes = chooser.Routes()
	}
	receiver := reflect.ValueOf(api)
	endpoints := make([]endpoint, 0, apiType.NumMethod())
	for index := 0; index < apiType.NumMethod(); index++ {
		method := apiType.Method(index)
		if method.Name == "Routes" {
			continue
		}
		if err := checkShape(method); err != nil {
			return nil, err
		}
		route, chosen := routes[method.Name]
		if !chosen {
			route = RouteOf(method.Name)
		}
		endpoints = append(endpoints, endpoint{
			name:       method.Name,
			route:      route,
			httpMethod: httpMethodOf(method),
			method:     method,
			call:       receiver.Method(index),
		})
	}
	return endpoints, nil
}

// checkShape says whether an endpoint can serve method. The receiver is the
// method's first parameter; the request, when there is one, is the second.
func checkShape(method reflect.Method) error {
	if method.Type.NumIn() > 2 {
		return fmt.Errorf("koba: method %s takes %d parameters; an endpoint supplies at most one, the request", method.Name, method.Type.NumIn()-1)
	}
	if method.Type.NumIn() == 2 && method.Type.In(1).Kind() == reflect.Pointer {
		return fmt.Errorf("koba: method %s takes a pointer; an endpoint decodes the request into a value", method.Name)
	}
	if method.Type.NumOut() > 1 {
		return fmt.Errorf("koba: method %s returns %d values; an endpoint responds with at most one", method.Name, method.Type.NumOut())
	}
	return nil
}

// httpMethodOf is the HTTP method an endpoint answers to: POST when the
// method takes a request, which arrives as the body, and GET when it does
// not.
func httpMethodOf(method reflect.Method) string {
	if method.Type.NumIn() > 1 {
		return http.MethodPost
	}
	return http.MethodGet
}

// RouteOf is the route an endpoint gets from the name of its method: the
// name in lower case, with a dash between words. "DeclarePoll" is served at
// "/declare-poll" and "ListHTTPStatuses" at "/list-http-statuses".
func RouteOf(methodName string) string {
	letters := []rune(methodName)
	var route strings.Builder
	route.WriteByte('/')
	for index, letter := range letters {
		if index > 0 && startsWord(letters, index) {
			route.WriteByte('-')
		}
		route.WriteRune(unicode.ToLower(letter))
	}
	return route.String()
}

// startsWord reports whether the upper-case letter at index begins a word:
// it follows a lower-case letter or a digit, or it is the last capital of an
// acronym before a lower-case letter.
func startsWord(letters []rune, index int) bool {
	if !unicode.IsUpper(letters[index]) {
		return false
	}
	previous := letters[index-1]
	if !unicode.IsUpper(previous) {
		return true
	}
	return index+1 < len(letters) && unicode.IsLower(letters[index+1])
}

// handler serves the endpoint: it decodes the request into the method's
// parameter, calls the method and writes what it returned as JSON, or 204 No
// Content when it returns nothing. A body that does not decode is answered
// with 400 Bad Request; a panic in the method is answered as a failure.
func (served endpoint) handler() httprouter.HandlerFunc {
	return func(context *httprouter.Context) {
		arguments, err := served.arguments(context)
		if err != nil {
			writeRejection(context, err)
			return
		}
		results, failure := invoke(served.call, arguments)
		if failure != nil {
			writeFailure(context, failure)
			return
		}
		writeResults(context, results)
	}
}

// arguments builds the values the method is called with, one per parameter
// after the receiver: the request, decoded from the JSON body.
func (served endpoint) arguments(context *httprouter.Context) ([]reflect.Value, error) {
	signature := served.method.Type
	arguments := make([]reflect.Value, 0, signature.NumIn()-1)
	for index := 1; index < signature.NumIn(); index++ {
		request := reflect.New(signature.In(index))
		if err := context.ShouldBindJSON(request.Interface()); err != nil {
			return nil, fmt.Errorf("request body: %w", err)
		}
		arguments = append(arguments, request.Elem())
	}
	return arguments, nil
}

// invoke calls the method and reports a panic as the failure instead of
// letting it escape.
func invoke(call reflect.Value, arguments []reflect.Value) (results []reflect.Value, failure any) {
	defer func() {
		if recovered := recover(); recovered != nil {
			failure = recovered
		}
	}()
	return call.Call(arguments), nil
}

func writeResults(context *httprouter.Context, results []reflect.Value) {
	if len(results) == 0 {
		context.Status(http.StatusNoContent)
		return
	}
	context.JSON(http.StatusOK, results[0].Interface())
}

// writeRejection answers a request the endpoint could not read.
func writeRejection(context *httprouter.Context, err error) {
	writeError(context, http.StatusBadRequest, err.Error())
}

// writeFailure answers a call that failed.
func writeFailure(context *httprouter.Context, failure any) {
	writeError(context, http.StatusInternalServerError, messageOf(failure))
}

// writeError answers with a JSON body that names the status and carries the
// message.
func writeError(context *httprouter.Context, status int, message string) {
	context.JSON(status, httprouter.H{
		"status":  status,
		"error":   statusName(status),
		"message": message,
	})
}

// statusName is the status in the form an error body carries: its standard
// text in lower case with underscores, "not_found" for 404.
func statusName(status int) string {
	return strings.ReplaceAll(strings.ToLower(http.StatusText(status)), " ", "_")
}

// messageOf is the text a failure is reported with: an error's Error, and
// anything else printed.
func messageOf(failure any) string {
	if err, ok := failure.(error); ok {
		return err.Error()
	}
	return fmt.Sprint(failure)
}
